src/Qir/Runtime/lib/Simulators/FullstateSimulator.cpp

```diff
@@ -272,71 +300,87 @@ namespace Quantum
         {
             static TSingleQubitGate op = reinterpret_cast<TSingleQubitGate>(this->GetProc("X"));
             op(this->simulatorId, GetQubitId(q));
+            UnmarkSingleQubit(q);
         }
 
         void ControlledX(long numControls, Qubit controls[], Qubit target) override
         {
             static TSingleQubitControlledGate op = reinterpret_cast<TSingleQubitControlledGate>(this->GetProc("MCX"));
             std::vector<unsigned> ids            = GetQubitIds(numControls, controls);
             op(this->simulatorId, (unsigned)numControls, ids.data(), GetQubitId(target));
+            UnmarkSingleQubit(target);
+            UnmarkQubitList(numControls, controls);

```

Comment:
Are you sure that we need to mark the controls as not measured upon the `Controlled..()` operations?

My impression was that controls stay unchanged during the `Controlled..()` operations (they do not collapse). If they are already measured, then they should still be considered measured after the `Controlled..()` operation.

Reply:
In general, controlled operations can affect the state of all qubits that they act on, whether those are controls or targets. As an example, phase kickback is a technique (used in iterative phase estimation and in the Deutsch–Jozsa problem) where the target stays in a fixed state, but the phase imparted by an operation call is associated with a control register instead.

Another way to see it is that these two Q# programs are equivalent when executed on ideal devices:

```qsharp
within {
    H(left);
    H(right);
} apply {
    CNOT(left, right);
}
// and
CNOT(right, left);
```

As a result, when applying a controlled operation to a register whose state is known exactly, it is no longer the case that the state is known unless one spends possibly exponential overhead to track the state.

Reply:
Deja vu ;) It wasn't so long ago that you were explaining this to me (https://github.com/microsoft/qsharp-runtime/pull/231#issuecomment-639585799)